alica_engine/include/engine/AlicaEngine.h

```diff
@@ -44,95 +33,82 @@ class AlicaEngine
-    TeamManager* getTeamManager() const { return _teamManager; }
-    TeamObserver* getTeamObserver() const { return _teamObserver; }
-    AlicaClock* getAlicaClock() const { return _alicaClock; }
+    AuthorityManager& getAuth() { return _auth; }
+    BehaviourPool& getBehaviourPool() { return _behaviourPool; }
+    Logger& getLog() { return _log; }
+    PlanBase& getPlanBase() { return _planBase; }
+    PlanParser& getPlanParser() { return _planParser; }
+    PlanRepository& getPlanRepository() { return _planRepository; }
+    VariableSyncModule& getResultStore() { return *_variableSyncModule; }
+    IRoleAssignment& getRoleAssignment() { return *_roleAssignment; }
+    SyncModule& getSyncModul() { return _syncModul; }
+    TeamManager& getTeamManager() { return _teamManager; }
+    TeamObserver& getTeamObserver() { return _teamObserver; }
+    AlicaClock& getAlicaClock() { return _alicaClock; }
 
     const BlackBoard& getBlackBoard() const { return _blackboard; }
     BlackBoard& editBlackBoard() { return _blackboard; }
-    // Solver Access:
-    template <class SolverType>
-    void addSolver(SolverType* solver);
-    template <class SolverType>
-    SolverType* getSolver() const;
 
     // Data Access:
-
     const RoleSet* getRoleSet() const { return _roleSet; }
 
-    // Setters:
-    void setMaySendMessages(bool maySendMessages);
-
-    // Module Setters TODO: remove and replace with a better configuration system
-    void setLog(Logger* log);
-    void setTeamObserver(TeamObserver* teamObserver);
-    void setSyncModul(SyncModule* syncModul);
-    void setAuth(AuthorityManager* auth);
-    void setRoleAssignment(IRoleAssignment* roleAssignment);
-    void setCommunicator(IAlicaCommunication* communicator);
-    void setAlicaClock(AlicaClock* clock);
-    void setResultStore(VariableSyncModule* resultStore);
-
     // internals
     void setStepCalled(bool stepCalled);
     bool getStepCalled() const;
     void iterationComplete();
 
-    // AgentIDManager forwarded interface:
+    // AlicaContext forward interface
+    const IAlicaCommunication& getCommunicator() const;
+    std::string getRobotName() const;
+    template <class SolverType>
+    SolverType& getSolver() const;
+    template <class SolverType>
+    bool existSolver() const;
 
-    AgentIDConstPtr getIdFromBytes(const std::vector<uint8_t>& vectorID) const;
+    // AgentIDManager forwarded interface:
+    AgentIDConstPtr getIdFromBytes(const std::vector<uint8_t>& vectorID);
     template <class Prototype>
-    AgentIDConstPtr getId(Prototype& idPrototype) const;
+    AgentIDConstPtr getId(Prototype& idPrototype);
 
 private:
     void setStepEngine(bool stepEngine);
 
-    PlanBase* _planBase;
-    TeamObserver* _teamObserver;
-    ExpressionHandler* _expressionHandler;
-    BehaviourPool* _behaviourPool;
+    AlicaContext& _ctx;
+    essentials::AgentIDManager _agentIDManager;
+    PlanRepository _planRepository;
+    PlanParser _planParser;
+    const Plan* _masterPlan;
```

Comment:
Well apparently there is dependency. Member are constructed in order they are declared here. I had to match the earlier initialization order to avoid weird crashes.

Reply:
True, but you should be able to safely move pointer members up

Reply:
not really, this is populated by call to parsePlanTree() which must come after planparser and planrepository construction and before others. Check contructor initializer list once.

Reply:
Yes, I see now. In this case, please mark these members with big warning comment that there is an initializer order dependency and that they cannot be moved freely.

Reply:
Aye